Implemented VTK/Silo field selection from input file

francoishamon opened the pull request:
For large cases, our default output VTK/Silo strategy produces big output files that are no easy to manipulate/share. This PR implements a way to select in the XML the fields that we want to include in the VTK/Silo file. The `Outputs` block looks like that:
```
  <Outputs>                                                                                                                                                                              
    <VTK                                                                                                                                                                                 
      name="vtkOutputPoromecaHyst"                                                                                                                                                       
      fieldNames="{ TotalDisplacement, pressure, phaseVolumeFraction, rockReservoir_stress, rockBurdens_stress }"/>                                                                      
  </Outputs>   
```

Rebaseline done in https://github.com/GEOSX/integratedTests/pull/235

TotoGaz commented on src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.hpp:
```diff
@@ -117,6 +117,16 @@ class VTKPolyDataWriterInterface
+   * @brief Set the names of the fields to output
+   * @param[in] fieldNames the fields to output
+   */
+  void setFieldNames( array1d< string > const & fieldNames )
```
```suggestion
  void setFieldNames( array1d< string const > const & fieldNames )
```

francoishamon replied:
Done

TotoGaz commented on src/coreComponents/fileIO/silo/SiloFile.hpp:
```diff
@@ -662,8 +662,26 @@ class SiloFile
+   * @brief Check if plotting is enabled for this field
+   * @param[in] wrapperPlotLevel the plot level of the wrapper
+   * @param[in] wrapperName the name of the wrapper
+   */
```
`@return`

francoishamon replied:
Done, I fixed the Doxygen here

TotoGaz commented on src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp:
```diff
@@ -912,5 +913,21 @@ void VTKPolyDataWriterInterface::write( real64 const time,
+bool VTKPolyDataWriterInterface::isFieldPlotEnabled( PlotLevel const wrapperPlotLevel,
+                                                     string const & wrapperName ) const
+{
+  // check if the logLevel is sufficient high for plotting
+  bool plotEnabled = ( wrapperPlotLevel <= m_plotLevel );
+
+  // override the logLevel if the fieldNames list was provided
+  if( !m_fieldNames.empty() )
+  {
+    auto search = m_fieldNames.find( wrapperName );
+    plotEnabled = ( search != m_fieldNames.end() );
+  }
+  return plotEnabled;
```
It looks like I've seen this code before 🤣

francoishamon replied:
Yes, I agree, this code is duplicated. It is not easy to do something better because they live in different folders (`vtk` and `silo`), but I will find something

francoishamon replied:
Done, this piece of code was indeed duplicated, but I removed the duplication in the latest commit

TotoGaz commented on src/coreComponents/fileIO/silo/SiloFile.cpp:
```diff
@@ -3167,5 +3167,21 @@ int SiloFile::getMeshType( string const & meshName ) const
   return meshType;
 }
 
+bool SiloFile::isFieldPlotEnabled( PlotLevel const wrapperPlotLevel,
```
You could add a
```c++
bool SiloFile::isFieldPlotEnabled( string WrapperBase const & wrapper ) const
{
    return this->isFieldPlotEnabled( wrapper.getPlotLevel(), wrapper.getName() );
}
```
which can bring some smoother reading.

francoishamon replied:
Done, I implemented what you suggested

TotoGaz commented on inputFiles/poromechanics/PoroElastic_deadoil_3ph_baker_2d.xml:
```diff
@@ -279,10 +279,12 @@
     <Silo
-      name="siloOutput"/>
+      name="siloOutput"
+      fieldNames="{ TotalDisplacement, pressure, skeleton_stress }"/>
```
If possible, I would definitely implement a `from` `to` (some kind of a mapping).
That could be very useful for our users.

francoishamon replied:
@TotoGaz what kind of mapping do you suggest? Is there a case in which we want the names that appear in the vtk files to be different from the `fieldNames` in GEOSX?

TotoGaz replied:
Oh, it's the output, not the input, stupid me 😵‍💫 
Forget what I said.

cssherman commented:
Thanks for working on this!  I had been tinkering with the same idea, but ran into some issues with my approach (temporarily toggling plot level flags before handing it off to the individual readers).

What are your thoughts on allowing the user to enter a path-like string as well, so that we can discriminate between similarly named fields in different regions?  Or is the current region filter enough sufficient?

francoishamon commented:
> Thanks for working on this! I had been tinkering with the same idea, but ran into some issues with my approach (temporarily toggling plot level flags before handing it off to the individual readers).
> 
> What are your thoughts on allowing the user to enter a path-like string as well, so that we can discriminate between similarly named fields in different regions? Or is the current region filter enough sufficient?

Yes, thanks for the suggestion, I agree that this could be an additional refinement. For my own usage, the current filter is sufficient because:
- For the material fields, I can use the name of the model as an additional filter if I don't want to output the quantity everywhere.
- For the regular fields, I typically need them on all the target regions of the solvers, so not having a region-filter is good enough.

We can discuss that at the next developer meeting.

CusiniM reviewed (approved):
I ll just throw the idea in there. Feel free to ignore it. 

Maybe, we should have a step to check whether the fields specified actually exist within GEOSX. It should be fairly easy to check if there is, at least, an `extrinsicMeshData` with that name. I guess it's a lot harder to know whether any of the solver specified has actually registered the required field. I am only saying this coz we kind of require the user to know that the variable pressure inside GEOSX is called `pressure` and not `Pressure` or `PRESSURE`.

francoishamon commented:
> I ll just throw the idea in there. Feel free to ignore it.
> 
> Maybe, we should have a step to check whether the fields specified actually exist within GEOSX. It should be fairly easy to check if there is, at least, an `extrinsicMeshData` with that name. I guess it's a lot harder to know whether any of the solver specified has actually registered the required field. I am only saying this coz we kind of require the user to know that the variable pressure inside GEOSX is called `pressure` and not `Pressure` or `PRESSURE`. Running a long test case and finding out at the end that the name of a field was wrong so the field one wanted to plot is wrong can be a bit frustrating.

Yes, good idea, I will give it a try, thanks!

cssherman commented:
This addresses #1844

francoishamon commented:
Following the comment from @CusiniM I added a check to make sure that the `fieldNames` specified by the user are indeed registered either on the `nodeManager` or on an element subregion. The check takes place the first time the vtk ouput is called, and issues a warning if a field names is not registered.

Following a comment from @rrsettgast during a dev meeting I added a flag that can be used to specify whether we only plot the `fieldNames` listed by the user in the xml, or the `fieldNames` *and* all the fields with the required plot level.
